bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
 use crate::migrate::{
     Device, MigrateError, MigratePhase, MigrateRole, MigrationState, PageIter,
 };
-use crate::vm::migrate_commands::MigrateTargetCommand;
-use crate::vm::migrate_commands::MigrateTargetResponse;
+use crate::vm::ensure::{VmEnsureActive, VmEnsureNotStarted};
+use crate::vm::objects::VmObjectsShared;
+use crate::vm::state_publisher::{ExternalStateUpdate, MigrationStateUpdate};
 
 use super::protocol::Protocol;
+use super::MigrateConn;
+
+/// The interface to an arbitrary version of the target half of the live
+/// migration protocol.
+//
+// Use `async_trait` here to help generate a `Send` bound on the futures
+// returned by the functions in this trait.
+#[async_trait::async_trait]
+pub(crate) trait DestinationProtocol<'e> {
```

Comment:
not 100% sure if I get why this lifetime parameter is necessary? but AFAICT it doesn't seem to be causing any downstream complexity, so :woman_shrugging:

Reply:
Embarrassingly, I got here mostly by throwing things at rustc until it compiled :grimacing: But, if I understand things right (and I very well may not), the issues here are:

- If there's no extra lifetime bound on the trait itself, then when we `impl DestinationProtocol for RonV0<'e>`, the compiler is unsure that the `'e` references in `self` won't be moved/copied into the `'ensure` return value. This produces a "lifetime may not live long enough" error with a hint that an extra `'e: 'ensure` bound would fix the problem.
- But adding `where 'e: 'ensure` just to this impl produces a stronger bound on the lifetimes in the impl than were expressed in the original trait, which is also a compiler error. The easiest way out of that is to name the extra lifetime in the trait and add the appropriate `where` bound to the trait definition, as I've done here.

With all that said: all this goes away if `RonV0` doesn't hold onto an `EnsureState` (as you suggested below), and I like the way that change is looking in my early hacking. If it holds up it should eliminate the need for this extra lifetime.

Reply:
The latest push did indeed resolve this by getting rid of the relevant `RonV0` member (and in fact by getting rid of `EnsureState` entirely).